fuel-vm/src/interpreter/blockchain.rs

```diff
@@ -650,15 +648,25 @@ where
         *self.sp = new_sp;
         *self.ssp = new_sp;
 
-        // Copy the code.
-        copy_from_slice_zero_fill(
-            self.memory,
-            owner,
-            contract_bytes,
-            region_start,
-            contract_offset,
-            length,
-        )?;
+        // Get the slice where to read the contract code into
+        let contract_buffer = self.memory.write(owner, region_start, length)?;
+
+        if contract_offset >= contract_len.into() {
+            contract_buffer[..].fill(0);
+        } else {
+            // The conract offset is within the contract bounds, which
+            // is guaranteed to be at most `u32::MAX`. As such, it is safe
+            // to convert `contract_offset` to `usize`, which is required
+            // to invoke `self.storage.read_contract`. Furthermore,
+            // the call to `self.storage.read_contract` cannot fail with
+            // error `StorageError::`
+            #[allow(clippy::cast_possible_truncation)]
+            self.storage
+                .read_contract(&contract_id, contract_offset as usize, contract_buffer)
```

Comment:
I think the `fuel-vm` should be responsible for extracting subsection from `contract_buffer` and passing it to the `read_contract` function.

Plus, it should be responsible for filling `contract_buffer[(length - (contract_len - contract_offset))..].fill(0);` as well.

Because it is requirement from the specification and we don't want to leak business logic to the storage.

The same is related to another place in the PR.